### ISSUE TITLE
padding-line-between-statements: Missing functionality covered by deprecated lines-around-directive

### ISSUE THREAD

**Issue by nwoltman** [eslint/eslint#11129]
**Tell us about your environment**

* **ESLint Version:** 5.9.0
* **Node Version:** 10.13.0
* **npm Version:** 6.4.1

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {
  rules: {
    'lines-around-directive': ['error', { before: 'never', after: 'always' }],

    'padding-line-between-statements': ['error',
      { blankLine: 'never', prev: '*', next: 'directive' },
      { blankLine: 'always', prev: 'directive', next: '*' },
    ],
  }
}
```
</details>
<br/>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
<empty line>
'use strict';

console.log(1);
```

also

```js
/**
 * Copyright
 */

'use strict';

console.log(1);
```

command:
```bash
eslint .
```

**What did you expect to happen?**

Both the [`lines-around-directive`](https://eslint.org/docs/rules/lines-around-directive) (deprecated) rule and [`padding-line-between-statements`](https://eslint.org/docs/rules/padding-line-between-statements) rules should warn about the unexpected empty line at the start of the file.

**What actually happened? Please include the actual, raw output from ESLint.**

Only the `lines-around-directive` rule produced an error:

```
  2:1  error  Unexpected newline before "use strict" directive  lines-around-directive
```

**Are you willing to submit a pull request to fix this bug?**

Maybe. The hard part here is figuring out what to do about this.

---

The problem seems to be that `prev: '*'` in the `padding-line-between-statements` configuration doesn't match the top of the file before the `'use strict';` directive.

This is a real problem because the `lines-around-directive` rule was deprecated in ESLint v4.0.0 with the claim that it will be replaced by the `padding-line-between-statements` rule. However, the `padding-line-between-statements` rule is not able to fully replicate the functionality of the `lines-around-directive` rule because of this bug.

**Comment by platinumazure** [comment]
Labeling this as bug and enhancement for now since it seems it could go either way.

I'll champion this (if it does turn into an enhancement). This is a backward compatibility problem as far as I'm concerned.

I think I would prefer having top or bottom of file be a "statement type" that isn't applied by default with `*` configurations, but open to bikeshedding on that.

**Comment by nwoltman** [comment]
If the `*` statement type also matched the top/bottom of files, that could create contradictions with other rules:

```js
{
  'eol-last': ['error', 'always'],

  'padding-line-between-statements': ['error',
    { blankLine: 'never', prev: 'cjs-export', next: '*' }
  ],
}
```

```js
module.exports.foo = 1;
module.exports.bar = 2;
<EOF>
```

`padding-line-between-statements` would raise an error here because it expects no newlines after an export, but `eol-last` expects a newline at the end of the file, so removing the newline would also be an error.

(Though I'm not sure why someone would use `next: '*'` in this case instead of a specific statement type.)

**Comment by kaicataldo** [comment]
If we start including "beginning of file" and "end of file" as types, it feels like this rule should be called something like `padding-line-around-statements`.

I would advocate for [`no-multiple-empty-lines`](https://eslint.org/docs/rules/no-multiple-empty-lines) to take on the responsibility of reporting this, given the `maxBOF` configuration option it has. If that rule doesn't work for this case, I think we should update that rule to cover this case instead.

**Comment by nwoltman** [comment]
The `no-multiple-empty-lines` rule currently doesn't help if there is a comment before the directive:

```js
/**
 * Copyright
 */
'use strict';
```

It only looks at empty lines before the comment (which IMO is the expected behavior).

**Comment by nwoltman** [comment]
It would be nice if we could figure out how to solve this for v6. Here's a summary of the discussion so far:

+ `padding-line-between-statements` maybe shouldn't have "beginning of file" and "end of file" types added since its job is to handle lines *between* statements. If these types were to be added, `padding-line-around-statements` would be a more appropriate name ([comment](https://github.com/eslint/eslint/issues/11129#issuecomment-442197336))
+ `no-multiple-empty-lines` already handles empty lines at the beginning of files, but it can't handle the case where there is a comment at the start of the file before the directive ([comment](https://github.com/eslint/eslint/issues/11129#issuecomment-442238978))

**Comment by IgorNovozhilov** [comment]
@nwoltman, @kaicataldo. I check this on v.6.4.0 `no-multiple-empty-lines` is worked now in this situation
example:
```js
    'no-multiple-empty-lines': ['warn',
      { max: 2, maxEOF: 0, maxBOF: 0 }
    ],
```

```
    1:1  warning  Too many blank lines at the beginning of file. Max of 0 allowed  no-multiple-empty-lines
  336:1  warning  Too many blank lines at the end of file. Max of 0 allowed        no-multiple-empty-lines

✖ 2 problems (0 errors, 2 warnings)
  0 errors and 2 warnings potentially fixable with the `--fix` option.
```

ps: `no-multiple-empty-lines` - Allows `\n` on the last line, and does not conflict with `eol-last`

**Comment by kaicataldo** [comment]
Can this issue be closed if the solution proposed by @IgorNovozhilov now works?

**Comment by nwoltman** [comment]
It doesn't look like anything is solved.

Config:

```json
{
  "root": true,
  "rules": {
    "lines-around-directive": ["error", {"before": "never", "after": "always"}],
    "no-multiple-empty-lines": ["error", {"max": 2, "maxEOF": 0, "maxBOF": 0}],
    "padding-line-between-statements": ["error",
      {"blankLine": "never", "prev": "*", "next": "directive"},
      {"blankLine": "always", "prev": "directive", "next": "*"}
    ]
  }
}
```

Code:

```js
/**
 * Copyright
 */

'use strict';

module.exports = 1;
```

ESLint output:

```
  5:1  error  Unexpected newline before "use strict" directive  lines-around-directive

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.
```

Only the `lines-around-directive` rule is creating an error. The behaviour for `no-multiple-empty-lines` is correct here anyway since the comment is on the first line, so there are no blank lines at the beginning of the file.

**Comment by IgorNovozhilov** [comment]
So... it seems, I all understood. 
I have 2 suggestions. But both are major changes that will lead to new warnings:
1. I could modify `lines-around-comment' rule so that it works with values for each of its options:
`true` - require a blank line
`false` - prohibits blank lines
`undefined` - validation is ignored
PS: But, I would like to mark this rule as deprecated, as rule `lines-around-directive`
```json
{
  "rules": {
    "lines-around-comment": ["error", {"beforeBlockComment": true, "afterBlockComment": false}],
    "no-multiple-empty-lines": ["error", {"max": 2, "maxEOF": 0, "maxBOF": 0}],
    "padding-line-between-statements": ["error",
      {"blankLine": "never", "prev": "*", "next": "directive"},
      {"blankLine": "always", "prev": "directive", "next": "*"}
    ]
  }
}
```
2. (me favorite). I could modify `padding-line-between-statements` rule by adding new type `comment`, it will work by the same rules as other types. And mark the rule `lines-around-comment' as deprecated 
```json
{
  "rules": {
    "no-multiple-empty-lines": ["error", {"max": 2, "maxEOF": 0, "maxBOF": 0}],
    "padding-line-between-statements": ["error",
      {"blankLine": "never", "prev": "comment", "next": "directive"},
      {"blankLine": "always", "prev": "directive", "next": "*"}
    ]
  }
}
```

**Comment by nwoltman** [comment]
Option 2 would be a nice solution 🙂. The only problem is that a _comment_ isn't a _statement_, so it doesn't exactly fit in the rule called `padding-line-between-statements`. Maybe we'd have to implement @kaicataldo's [suggestion](https://github.com/eslint/eslint/issues/11129#issuecomment-442197336) and call the rule `padding-line-around-statements`.

**Comment by IgorNovozhilov** [comment]
https://github.com/eslint/rfcs/pull/41

**Comment by kaicataldo** [comment]
One other thought: we currently have `padded-block`, `lines-around-comment`, `lines-between-class-members`, and `padding-line-around-statements`. 

Could we deprecate `padding-line-around-statements`, `lines-around-comment`, and `lines-between-class-members` in favor of a new rule `padding-lines` that allows for the enforcement of padding lines between any syntax in the same scope (including statements, properties, methods, comments, directives - everything!). It could essentially have the same API and configuration options it does now with the addition of options needed to cover the cases from the other rules.

We get then deprecate `padded-block` in favor of a new rule called `padded-scope` that covers all that `padded-block` currently does, but also cover the file/module scope (and therefore cover `BOF` and `EOF`).

This way, we could have two padding line rules - one for padding lines between syntax in the same scope and one for padding lines within new scopes (including the file/module scope).

I know our padding lines rules have changed a lot over the years, but I don't know how useful it is to have multiple rules for different syntax, and I think having two more generic, highly configurable rules could allow for users to have more control over this.

If the team is amenable to this, I can do some more research and write up a formal proposal.

**Comment by IgorNovozhilov** [comment]
I think at the moment you can solve this problem by this way: https://github.com/eslint/eslint/pull/12342, since this is a solution with minimal changes, whereas in future versions consider proposals to restructure the rules handling empty lines

**Comment by mdjermanovic** [comment]
> Could we deprecate `padding-line-around-statements`, `lines-around-comment`, and `lines-between-class-members` in favor of a new rule `padding-lines` that allows for the enforcement of padding lines between any syntax in the same scope (including statements, properties, methods, comments, directives - everything!). 

This seems like a good idea since the `lines-around-comment` can now clash with the other two rules, like in this [Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IFwicGFkZGluZy1saW5lLWJldHdlZW4tc3RhdGVtZW50c1wiOlxuICAgICBbXG4gICAgICAgIFwiZXJyb3JcIixcbiAgICAgICAgeyBcImJsYW5rTGluZVwiOiBcIm5ldmVyXCIsIFwicHJldlwiOiBcImNvbnN0XCIsIFwibmV4dFwiOiBcImxldFwiIH0sXG4gICAgIF1cbiovXG4vKiBlc2xpbnQgbGluZXMtYXJvdW5kLWNvbW1lbnQ6IFtcImVycm9yXCIsIHsgXCJiZWZvcmVMaW5lQ29tbWVudFwiOiB0cnVlIH1dKi9cblxuY29uc3QgYSA9IDE7XG4vLyBjb21tZW50XG5sZXQgYjsiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjYsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsiY29uc3RydWN0b3Itc3VwZXIiOjIsImZvci1kaXJlY3Rpb24iOjIsImdldHRlci1yZXR1cm4iOjIsIm5vLWFzeW5jLXByb21pc2UtZXhlY3V0b3IiOjIsIm5vLWNhc2UtZGVjbGFyYXRpb25zIjoyLCJuby1jbGFzcy1hc3NpZ24iOjIsIm5vLWNvbXBhcmUtbmVnLXplcm8iOjIsIm5vLWNvbmQtYXNzaWduIjoyLCJuby1jb25zdC1hc3NpZ24iOjIsIm5vLWNvbnN0YW50LWNvbmRpdGlvbiI6Miwibm8tY29udHJvbC1yZWdleCI6Miwibm8tZGVidWdnZXIiOjIsIm5vLWRlbGV0ZS12YXIiOjIsIm5vLWR1cGUtYXJncyI6Miwibm8tZHVwZS1jbGFzcy1tZW1iZXJzIjoyLCJuby1kdXBlLWtleXMiOjIsIm5vLWR1cGxpY2F0ZS1jYXNlIjoyLCJuby1lbXB0eSI6Miwibm8tZW1wdHktY2hhcmFjdGVyLWNsYXNzIjoyLCJuby1lbXB0eS1wYXR0ZXJuIjoyLCJuby1leC1hc3NpZ24iOjIsIm5vLWV4dHJhLWJvb2xlYW4tY2FzdCI6Miwibm8tZXh0cmEtc2VtaSI6Miwibm8tZmFsbHRocm91Z2giOjIsIm5vLWZ1bmMtYXNzaWduIjoyLCJuby1nbG9iYWwtYXNzaWduIjoyLCJuby1pbm5lci1kZWNsYXJhdGlvbnMiOjIsIm5vLWludmFsaWQtcmVnZXhwIjoyLCJuby1pcnJlZ3VsYXItd2hpdGVzcGFjZSI6Miwibm8tbWlzbGVhZGluZy1jaGFyYWN0ZXItY2xhc3MiOjIsIm5vLW1peGVkLXNwYWNlcy1hbmQtdGFicyI6Miwibm8tbmV3LXN5bWJvbCI6Miwibm8tb2JqLWNhbGxzIjoyLCJuby1vY3RhbCI6Miwibm8tcHJvdG90eXBlLWJ1aWx0aW5zIjoyLCJuby1yZWRlY2xhcmUiOjIsIm5vLXJlZ2V4LXNwYWNlcyI6Miwibm8tc2VsZi1hc3NpZ24iOjIsIm5vLXNoYWRvdy1yZXN0cmljdGVkLW5hbWVzIjoyLCJuby1zcGFyc2UtYXJyYXlzIjoyLCJuby10aGlzLWJlZm9yZS1zdXBlciI6Miwibm8tdW5kZWYiOjIsIm5vLXVuZXhwZWN0ZWQtbXVsdGlsaW5lIjoyLCJuby11bnJlYWNoYWJsZSI6Miwibm8tdW5zYWZlLWZpbmFsbHkiOjIsIm5vLXVuc2FmZS1uZWdhdGlvbiI6Miwibm8tdW51c2VkLWxhYmVscyI6Miwibm8tdXNlbGVzcy1jYXRjaCI6Miwibm8tdXNlbGVzcy1lc2NhcGUiOjIsIm5vLXdpdGgiOjIsInJlcXVpcmUtYXRvbWljLXVwZGF0ZXMiOjIsInJlcXVpcmUteWllbGQiOjIsInVzZS1pc25hbiI6MiwidmFsaWQtdHlwZW9mIjoyfSwiZW52Ijp7fX19)

**Comment by tomm1996** [comment]
So the fix would be to conflate `padding-line-around-statements`, `lines-around-comment` and `lines-between-class-members` into a more configurable `padding-lines`?

**Comment by nzakas** [comment]
Closing due to age.
